lib/matplotlib/_pylab_helpers.py

```diff
@@ -41,90 +44,73 @@ def get_fig_manager(cls, num):
     @classmethod
     def destroy(cls, num):
         """
-        Try to remove all traces of figure *num*.
+        Destroy figure number *num*.
 
-        In the interactive backends, this is bound to the
-        window "destroy" and "delete" events.
+        In the interactive backends, this is bound to the window "destroy" and
+        "delete" events.
         """
         if not cls.has_fignum(num):
             return
-        manager = cls.figs[num]
+        manager = cls.figs.pop(num)
         manager.canvas.mpl_disconnect(manager._cidgcf)
-        cls._activeQue.remove(manager)
-        del cls.figs[num]
         manager.destroy()
         gc.collect(1)
 
     @classmethod
     def destroy_fig(cls, fig):
-        "*fig* is a Figure instance"
-        num = next((manager.num for manager in cls.figs.values()
-                    if manager.canvas.figure == fig), None)
-        if num is not None:
-            cls.destroy(num)
+        """Destroy figure *fig*."""
+        canvas = getattr(fig, "canvas", None)
+        manager = getattr(canvas, "manager", None)
+        num = getattr(manager, "num", None)
```

Comment:
Is this always equivalent, since it seems to be working in reverse?

Reply:
This basically gets `fig.canvas.manager.num` with the provision that each of these can be None.  If this num is not registered in the `figs` dict, `destroy` will catch that as it first checks `cls.has_fignum(num)`.
It's true that this may cause problems if two managers share the same num, but then the num->manager mapping is already invalid (`set_active` will already have overwritten the entry in the dict) so I'm not too worried about that.